Added bulk update with single query

thisago opened the pull request:
~~For now, just SQLite module is updating with single query, soon I'll implement in Postgres~~

Now I added bulk update for SQLite and PostgreSQL, The tests are running well

Closes #188

PhilippMDoerner commented on changelog.md:
```diff
@@ -7,6 +7,9 @@
 -   [t]—test suite improvement
 -   [d]—docs improvement
 
+## 2.7.2 (March 24, 2023)
```
This would add a feature and thus turn it into 2.8.0
I'd change the 2.7.1 below to 2.8.0 and move the line to down there.
It's XX because XX is the day of the release, not the day of the merge or commit

thisago commented on norm.nimble:
```diff
@@ -1,6 +1,6 @@
 # Package
 
-version       = "2.7.0"
+version       = "2.7.2"
```
Do you want me to edit?

moigagoo replied:
Yes please. Add your changes under the latest WIP entry.

When time comes to cut the release, we'll determine the version (2.7.1 or 2.8.0, but I'm voting 2.8.0), and fix the release date.

moigagoo commented on src/norm/postgres.nim:
```diff
@@ -393,8 +393,57 @@ proc update*[T: Model](dbConn; obj: var T) =
+  # for obj in objs.mitems:
+  #   dbConn.update(obj)
+
+  # return
```
@thisago Please clean up commented lines.

moigagoo commented on src/norm/postgres.nim:
```diff
@@ -393,8 +393,57 @@ proc update*[T: Model](dbConn; obj: var T) =
+
+  type BulkUpdateValues = Table[string, seq[tuple[id: int64; val: DbValue]]]
+
+  func genBulkUpdateQuery(tableName: string; values: BulkUpdateValues): tuple[qry: string; values: seq[DbValue]] =
```
Please add doc comments to your new procs and types.

moigagoo commented on src/norm/postgres.nim:
```diff
@@ -393,8 +393,57 @@ proc update*[T: Model](dbConn; obj: var T) =
+    var ids: seq[int64]
+    let limit = values.len - 1
+    var
+      i = -1
```
This is a weird initial value as the first thing you do with it is `inc` it.

Why not start with 0 and inc it at the end of the loop?

moigagoo commented on src/norm/postgres.nim:
```diff
@@ -393,8 +393,57 @@ proc update*[T: Model](dbConn; obj: var T) =
+    let limit = values.len - 1
+    var
+      i = -1
+      v = 1
```
`v` is an unclear variable name. This isn't Go, no need to keep things short to the degree of them becoming cryptic :-)

I mean, `i` is one thing, it's totally fine. But `v`, what does it stand for?

moigagoo commented on src/norm/sqlite.nim:
```diff
@@ -354,11 +354,50 @@ proc update*[T: Model](dbConn; obj: var T) =
   ## Update rows for each `Model`_ instance in open array.
+  type BulkUpdateValues = Table[string, seq[tuple[id: int64; val: DbValue]]]
+
+  func genBulkUpdateQuery(tableName: string; values: BulkUpdateValues): tuple[qry: string; values: seq[DbValue]] =
```
Is that a func definition inside `update`? Please make it a separate func definition in the outer scope.

moigagoo reviewed (changes requested):
I guess it's a WIP for now, amirite?

thisago commented:
I think that I fixed all the requested changes, in the last commit I made:
- moved changelog change to next release instead of making another version
- added comments to new funcs and types
- moved it to outer scope
- renamed `v` counter var to `argNum`
- now the `i` var starts in 0 (not -1 anymore)

thisago commented:
> I guess it's a WIP for now, amirite?

I'm already using it and looks good for me, the requested changes was done

moigagoo commented on src/norm/postgres.nim:
```diff
@@ -390,43 +390,42 @@ proc update*[T: Model](dbConn; obj: var T) =
+  for field, conds in values.pairs:
+    result.qry.add fmt"  {field} = CASE{'\l'}"
+    for (id, val) in conds:
+      result.qry.add fmt"    WHEN id = {id} THEN ${argNum}{'\l'}"
```
What's the purpose of those tabs in front of WHEN?

thisago replied:
Just pretty formatting, I'll remove

moigagoo commented on src/norm/postgres.nim:
```diff
@@ -390,43 +390,42 @@ proc update*[T: Model](dbConn; obj: var T) =
+    inc i
+  let idsStr = ids.join ", "
+  result.qry.add fmt"WHERE id IN ({idsStr})"
+proc add(t: var BulkUpdateValues; field: string; id: int64, value: DbValue) =
```
Please separate procs with blank lines.

In general, please make your code more sparse. It's much easier to read this way.

thisago replied:
ok

moigagoo reviewed (changes requested):
I've reviewed the code. Looks fine but we need tests to prove that.

I guess this is a rare case when we don't need new tests because this is already covered in trows.

However, I wish we had benchmarks. I know it's too much to ask for this PR and I'm ready to merge this one without benchmarks (just fix the formatting) but if you'd be interested to adding benchmarks that illustrate the performance boost that would be great.

thisago commented:
> but if you'd be interested to adding benchmarks that illustrate the performance boost that would be great.

haha, ok, there's a better way to do the benchmark or just run it with `time`?

moigagoo commented:
> > but if you'd be interested to adding benchmarks that illustrate the performance boost that would be great.
> 
> haha, ok, there's a better way to do the benchmark or just run it with `time`?

I was planning to use benchy for it.

The way I see it:
1. Create `benchmarks` folder.
2. Add benchy to norm.nimble.
3. Add `bench` task to norm.nimble.
4. Add a benchmark that updates 1000 rows

thisago commented:
I ran 2 times the bench 1x for each mode and the result is very satisfactory:

1th
```
min time             avg time             std dv   runs  name
      133.105 ms       133.105 ms  ±0.000  x1     Bulk update
139891.469 ms 139891.469 ms  ±0.000  x1     Update each row
```

2th
```
min time             avg time             std dv   runs  name
      129.890 ms       129.890 ms  ±0.000  x1    Bulk update
138162.341 ms 138162.341 ms  ±0.000  x1    Update each row
```

thisago commented:
I updated the benchmark code to generate the DB just one time and copy into another file to modify.  
I added also an indicator of DB generation progress, it shows a dot for every added row.

Also, an important thing to add next is bulk insert because insert 1k of simple rows are too slow...

Another benchmark result:
```
min time      avg time       std dv  runs name
   182.585 ms    182.585 ms  ±0.000  x1   Bulk update
174604.512 ms 174604.512 ms  ±0.000  x1   Update each row
```

moigagoo commented:
I'll do some minor fixes myself later and cut a new release.

Thanks for the great work!
